finetuner/data.py

```diff
@@ -0,0 +1,188 @@
+import csv
+import os
+from contextlib import nullcontext
+from dataclasses import dataclass
+from typing import Generator, Optional, TextIO, Tuple, Union
+
+from _finetuner.runner.stubs.model import get_stub
+from docarray import Document, DocumentArray
+from docarray.document.generators import _subsample
+from docarray.document.mixins.helper import _is_uri
+from genericpath import isfile
+
+from finetuner.constants import DEFAULT_TAG_KEY
+
+
+@dataclass
+class CSVOptions:
+    """Class containing options for reading CSV files
+
+    :param size: The number of rows that will be sampled.
+    :param sampling_rate: The sampling rate between [0, 1] indicating how many lines of
+        the CSV are skipped. a sampling rate of 1 means that none are skipped, 0.5
+        means that half are skipped, and 0 means that all lines are skipped.
+    :param dialect: A description of the expected format of the CSV, can either be an
+        object of the :class:`csv.Dialect` class, or one of the strings returned by the
+        :meth:`csv.list_dialects()' function.
+    :param encoding: The encoding of the CSV file.
+    :param is_labeled: Whether the second column of the CSV represents a label that
+        should be assigned to the item in the first column (True), or if it is another
+        item that should be semantically close to the first (False).
+    :param convert_to_blob: Whether uris to local files should be converted to blobs
+    """
+
+    size: Optional[int] = None
+    sampling_rate: Optional[float] = None
+    dialect: Union[str, 'csv.Dialect'] = 'auto'
+    encoding: str = 'utf-8'
+    is_labeled: bool = False
+    convert_to_blob: bool = True
+
+
+def build_dataset(
+    data: Union[str, TextIO, DocumentArray],
+    model: str,
+    csv_options: Optional[CSVOptions] = None,
+) -> Union[str, DocumentArray]:
+
+    if isinstance(data, (TextIO)) or (isinstance(data, str) and isfile(data)):
+        model_stub = get_stub(model, select_model='clip-text')
+        data = DocumentArray(
+            load_finetune_data_from_csv(
+                file=data,
+                task=model_stub.task,
+                options=csv_options or CSVOptions(),
+            )
+        )
+
+    return data
+
+
+def load_finetune_data_from_csv(
+    file: Union[str, TextIO],
+    task: str = 'text-to-text',
+    options: CSVOptions = CSVOptions(),
```

Comment:
I think it is not recommended to call a constructor in a function declaration, since it is then only called once and every function while use the exact same instance.
So it is better to do: 
```python
def load_finetune_data_from_csv(
[...]
    options=None,
[...]
):
  options = options or CSVOptions()